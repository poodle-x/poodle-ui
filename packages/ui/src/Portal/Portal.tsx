import React from "react";
import ReactDOM from "react-dom";
import useSafeLayoutEffect from "../hooks/useSafeLayoutEffect";
import setRef from "../utils/setRef";

export interface PortalProps {
	/**
	 * Mount the children to the custom target rather than append to the body.
	 */
	mountOn?: HTMLElement;
	children: React.ReactNode;
	/**
	 * Chang container tag
	 */
	portalElementTag?: string;
	/**
	 * Change portal name in container **data-portal** attribute
	 */
	portalDataName?: string;
	/**
	 * Mount the children to the target when equal true
	 */
	mount?: boolean;
}

export interface PortalContextValue {
	defaultMount?: HTMLElement | null;
}

export const PortalContext = React.createContext<
	PortalContextValue | undefined
>(undefined);

export const Portal: React.ForwardRefExoticComponent<
	React.PropsWithoutRef<PortalProps> & React.RefAttributes<HTMLElement>
> = React.forwardRef<HTMLElement, PortalProps>(
	(
		{
			portalDataName = "poodle-ui",
			portalElementTag = "div",
			mount,
			mountOn,
			children,
		},
		ref
	) => {
		const portalContext = React.useContext(PortalContext);

		const [portal, setPortal] = React.useState<{
			element: HTMLElement | null;
		}>({
			element: null,
		});

		useSafeLayoutEffect(() => {
			if (mount) {
				if (!mountOn && !portalContext?.defaultMount && !portal.element) {
					const newPortal = document.createElement(portalElementTag);
					newPortal.setAttribute("data-portal", portalDataName);
					document.body.appendChild(newPortal);
					setPortal({
						element: newPortal,
					});
					setRef(ref, newPortal);
				} else {
					if (mountOn || (portalContext && portalContext?.defaultMount)) {
						setRef(ref, mountOn || portalContext?.defaultMount || null);
					}
				}
			}
		}, [
			mount,
			mountOn,
			portal.element,
			portalContext,
			portalContext?.defaultMount,
			portalDataName,
			portalElementTag,
			ref,
		]);

		useSafeLayoutEffect(() => {
			if (!mount) {
				if (portal.element) {
					portal.element.parentNode?.removeChild(portal.element);
					setPortal({
						element: null,
					});
				}
				setRef(ref, null);
			}
		}, [mount, portal.element, ref]);

		return mount
			? ReactDOM.createPortal(
					children,
					mountOn ||
						portalContext?.defaultMount ||
						portal.element ||
						document.body
			  )
			: null;
	}
);

Portal.displayName = "Portal";

export default Portal;
