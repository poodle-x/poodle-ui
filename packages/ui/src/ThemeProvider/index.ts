import { ThemeProvider } from "./ThemeProvider";

export * from "./ThemeProvider";

export default ThemeProvider;
