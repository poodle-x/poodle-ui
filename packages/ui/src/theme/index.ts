import React from "react";
import {
	ThemeValue,
	SystemThemeConfig,
	BreakpointsThemeConfig,
	SpacingThemeConfig,
	ColorsThemeConfig,
	FontsThemeConfig,
	BordersThemeConfig,
	ShadowThemeConfig,
	LayoutsThemeConfig,
} from "@poodle/system";

import { CSSObjectSystem, Stylesheet } from "../styled";
import { BoxThemeConfig } from "../Box";
import { ButtonThemeConfig } from "../Button";
import { ColumnsThemeConfig } from "../Columns";
import { ColumnThemeConfig } from "../Column";
import { ContainerThemeConfig } from "../Container";
import { TableThemeConfig } from "../Table";
import { TableHeadThemeConfig } from "../TableHead";
import { TableBodyThemeConfig } from "../TableBody";
import { TableRowThemeConfig } from "../TableRow";
import { TableCellThemeConfig } from "../TableCell";
import { VisuallyHiddenThemeConfig } from "../VisuallyHidden";
import { CheckControlThemeConfig } from "../CheckControl";
import { CheckboxThemeConfig } from "../Checkbox";
import { RadioThemeConfig } from "../Radio";
import { InputThemeConfig } from "../Input";
import { InputAdornmentThemeConfig } from "../InputAdornment";
import { IconButtonThemeConfig } from "../IconButton";
import { ModalThemeConfig } from "../Modal";
import { ModalHeaderThemeConfig } from "../ModalHeader";
import { ModalBodyThemeConfig } from "../ModalBody";
import { ModalFooterThemeConfig } from "../ModalFooter";
import { SelectThemeConfig } from "../Select";
import { TabsThemeConfig } from "../Tabs";
import { TabThemeConfig } from "../Tab";
import { TabListThemeConfig } from "../TabList";
import { TabPanelThemeConfig } from "../TabPanel";

export { Stylesheet };

export { useTheme } from "./useTheme";

export type DirectionThemeConfig = "rtl" | "ltr";

export interface GlobalThemeConfig {
	styles?: ThemeValue<CSSObjectSystem>;
}

export type IconType = React.FC<
	React.PropsWithoutRef<React.SVGProps<SVGSVGElement>> &
		React.RefAttributes<SVGSVGElement>
>;

export interface BaseIcons {
	RadioButton?: IconType;
	RadioCheckedButton?: IconType;
	CheckboxButton?: IconType;
	CheckboxCheckedButton?: IconType;
	CheckboxIndeterminateButton?: IconType;
	ArrowDown?: IconType;
}

export interface ThemeConfig extends SystemThemeConfig {
	global?: GlobalThemeConfig;
	direction?: ThemeValue<DirectionThemeConfig>;
	baseIcons?: BaseIcons;
	Box?: BoxThemeConfig;
	Button?: ButtonThemeConfig;
	Columns?: ColumnsThemeConfig;
	Column?: ColumnThemeConfig;
	Container?: ContainerThemeConfig;
	Table?: TableThemeConfig;
	TableHead?: TableHeadThemeConfig;
	TableBody?: TableBodyThemeConfig;
	TableRow?: TableRowThemeConfig;
	TableCell?: TableCellThemeConfig;
	VisuallyHidden?: VisuallyHiddenThemeConfig;
	CheckControl?: CheckControlThemeConfig;
	Checkbox?: CheckboxThemeConfig;
	Radio?: RadioThemeConfig;
	Input?: InputThemeConfig;
	InputAdornment?: InputAdornmentThemeConfig;
	IconButton?: IconButtonThemeConfig;
	Modal?: ModalThemeConfig;
	ModalHeader?: ModalHeaderThemeConfig;
	ModalBody?: ModalBodyThemeConfig;
	ModalFooter?: ModalFooterThemeConfig;
	Select?: SelectThemeConfig;
	Tabs?: TabsThemeConfig;
	Tab?: TabThemeConfig;
	TabList?: TabListThemeConfig;
	TabPanel?: TabPanelThemeConfig;
}

export const defaultGlobal = (
	overrideTheme: ThemeConfig
): GlobalThemeConfig => {
	return {
		styles: overrideTheme.global?.styles,
	};
};

export const defaultBreakpoints = (
	overrideTheme: ThemeConfig
): BreakpointsThemeConfig => {
	return {
		scale: ["sm", "md", "lg", "xl"],
		...overrideTheme.breakpoints,
		sets: {
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
			xxl: 1400,
			...overrideTheme.breakpoints?.sets,
		},
	};
};

export const defaultSpacing = (
	overrideTheme: ThemeConfig
): SpacingThemeConfig => {
	return {
		scale: 4,
		...overrideTheme.spacing,
		sets: {
			s: 4,
			m: 8,
			l: 16,
			...overrideTheme.spacing?.sets,
		},
	};
};

export const defaultFonts = (overrideTheme: ThemeConfig): FontsThemeConfig => {
	return {
		scale: "1rem",
		...overrideTheme.fonts,
		base: {
			fontFamily:
				'-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
			fontStyle: "normal",
			fontWeight: "normal",
			lineHeight: "125%",
			fontSize: "1rem",
			...overrideTheme.fonts?.base,
		},
		weights: {
			normal: 400,
			semibold: 600,
			bold: 700,
			...overrideTheme.fonts?.weights,
		},
		sets: {
			input: {
				fontSize: "1rem",
				fontWeight: "normal",
				lineHeight: "125%",
			},
			button: {
				fontSize: "1rem",
				fontWeight: "bold",
				lineHeight: "150%",
			},
			label: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			helper: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelMedium: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1rem",
				lineHeight: "110%",
			},
			checkControlLabelSmall: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "0.875rem",
				lineHeight: "110%",
			},
			checkControlLabelLarge: {
				fontStyle: "normal",
				fontWeight: "normal",
				fontSize: "1.125rem",
				lineHeight: "110%",
			},
			modalTitle: {
				fontStyle: "normal",
				fontWeight: "bold",
				fontSize: "1.5rem",
				lineHeight: "133.33%",
			},
			...overrideTheme.fonts?.sets,
		},
	};
};

export const defaultColor = (overrideTheme: ThemeConfig): ColorsThemeConfig => {
	return {
		modes: {
			...overrideTheme.colors?.modes,
		},
		sets: {
			text: "#282828",
			disabledText: "#A6A6A6",

			primary: {
				light: "#D1EBFF",
				base: "#00478A",
				dark: "#003A70",
				textOnLight: "#00478A",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			negative: {
				base: "#AF1308",
				dark: "#8D1007",
				light: "#FCF3F3",
				textOnLight: "#AF1308",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			positive: {
				base: "#006122",
				dark: "#004D1B",
				light: "#E5FFEE",
				textOnLight: "#006122",
				textOnBase: "#fff",
				textOnDark: "#fff",
			},

			warn: {
				base: "#FFBB33",
				dark: "#FAAB00",
				light: "#FFF1DB",
				textOnLight: "#282828",
				textOnBase: "#282828",
				textOnDark: "#282828",
			},

			bg: {
				base: "#F8F8F8",
				100: "#E0E0E0",
				textOn100: "#282828",
				200: "#CCCCCC",
				textOn200: "#282828",
			},

			modal: {
				base: "#F8F8F8",
				footer: "#F0F0F0",
			},

			border: {
				base: "#CCCCCC",
			},

			...overrideTheme.colors?.sets,
		},
	};
};

export const defaultBorder = (
	overrideTheme: ThemeConfig
): BordersThemeConfig => {
	return {
		sets: {
			base: {
				style: "solid",
				width: "1px",
				color: "border",
			},
			...overrideTheme?.borders?.sets,
		},
		radius: {
			base: "4px",
			...overrideTheme?.borders?.radius,
		},
	};
};

export const defaultBoxShadow = (
	overrideTheme: ThemeConfig
): ShadowThemeConfig => {
	return {
		boxShadowSets: {
			lv1: "1px 2px 2px -1px rgba(0,0,0,0.4)",
			lv2: "2px 2px 5px 0px rgba(0,0,0,0.4)",
			lv3: "4px 4px 5px 0px rgba(0,0,0,0.4)",
			...overrideTheme?.shadows?.boxShadowSets,
		},
	};
};

export const defaultLayouts = (
	overrideTheme: ThemeConfig
): LayoutsThemeConfig => {
	return {
		zIndices: {
			modal: 2000,
			...overrideTheme?.layouts?.zIndices,
		},
	};
};

export function createTheme(overrideTheme: ThemeConfig = {}): ThemeConfig {
	return {
		direction: "ltr",
		...overrideTheme,
		global: defaultGlobal(overrideTheme),
		spacing: defaultSpacing(overrideTheme),
		breakpoints: defaultBreakpoints(overrideTheme),
		colors: defaultColor(overrideTheme),
		fonts: defaultFonts(overrideTheme),
		borders: defaultBorder(overrideTheme),
		shadows: defaultBoxShadow(overrideTheme),
		layouts: defaultLayouts(overrideTheme),
	};
}

export type StylesConfig<T extends string> = {
	[K in T]: ThemeValue<CSSObjectSystem>;
};

export interface StandardComponentProps {
	variant?: string;
	themeExtend?: ThemeConfig;
	theme?: ThemeConfig;
	sizeStyle?: string;
	colorStyle?: string;
}

export interface StandardThemeConfig<
	T extends string,
	Props extends { [key: string]: any }
> {
	defaultProps?: Partial<Props>;
	styles?: Partial<StylesConfig<T>>;
	variants?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	sizes?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	colorStyles?: {
		[key: string]: Partial<StylesConfig<T>>;
	};
	modes?: {
		[key: string]: {
			styles?: Partial<StylesConfig<T>>;
			variants?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			sizes?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
			colorStyles?: {
				[key: string]: Partial<StylesConfig<T>>;
			};
		};
	};
	overrides?: Partial<StylesConfig<T>>;
	overrideClasses?: Partial<
		{
			[K in T]: boolean | string;
		}
	>;
}

export * from "./system";
