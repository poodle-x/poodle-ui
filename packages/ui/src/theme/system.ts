import {
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	compose,
	createSystem,
	getMode,
	FontsProps,
	ColorsProps,
	SpacingProps,
	LayoutProps,
	BordersProps,
	bordersSystem,
	flexSystem,
	FlexProps,
	pseudosSystem,
	PseudosProps,
	shadowSystem,
	ShadowProps,
	GridProps,
	gridSystem,
	ListProps,
	listSystem,
	SystemConfigList,
	ResponsiveProp,
} from "@poodle/system";
import { CSSSystem } from "../styled";
import { StandardThemeConfig } from "./index";
import * as CSS from "csstype";

const additionalSystem: SystemConfigList = {
	objectFit: true,
	objectPosition: true,
	transition: true,
	cursor: true,
	pointerEvents: true,
};

export interface AdditionalSystemProps {
	/**
	 * The CSS `object-fit` property
	 */
	objectFit?: ResponsiveProp<CSS.Property.ObjectFit>;
	/**
	 * The CSS `object-fit` property
	 */
	objectPosition?: ResponsiveProp<CSS.Property.ObjectPosition>;
	/**
	 * The CSS `transition` property
	 */
	transition?: ResponsiveProp<CSS.Property.Transition>;
	/**
	 * The CSS `cursor` property
	 */
	cursor?: ResponsiveProp<CSS.Property.Cursor>;
	/**
	 * The CSS `cursor` property
	 */
	pointerEvents?: ResponsiveProp<CSS.Property.PointerEvents>;
}

const allSystem = compose(
	colorsSystem,
	fontsSystem,
	layoutSystem,
	spacingSystem,
	bordersSystem,
	flexSystem,
	pseudosSystem,
	shadowSystem,
	gridSystem,
	listSystem,
	additionalSystem
);

export const system = createSystem(allSystem);

interface CssSystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		GridProps,
		ListProps {}

export interface SystemProps
	extends SpacingProps,
		LayoutProps,
		FontsProps,
		ColorsProps,
		BordersProps,
		ShadowProps,
		FlexProps,
		PseudosProps<CssSystemProps>,
		GridProps,
		ListProps,
		AdditionalSystemProps {}

const additionalProps = [
	"textStyle",
	"variant",
	"themeExtend",
	"theme",
	"sizeStyle",
	"colorStyle",
];

export function removeSystemProps(props: { [key: string]: any }) {
	let resultProps = {};
	Object.keys(props).forEach((prop) => {
		let found = false;
		for (const name in allSystem) {
			if (prop === name || additionalProps.indexOf(prop) >= 0) {
				found = true;
				break;
			}
		}

		if (!found && ["sx"].indexOf(prop) === -1) {
			resultProps = {
				...resultProps,
				[prop]: props[prop],
			};
		}
	});

	return resultProps;
}

export function createCSSSystemStandard<Key extends string>(data: {
	key: Key;
	config?: StandardThemeConfig<Key, any>;
	props: { [key: string]: any };
	base?: CSSSystem["base"];
}): CSSSystem {
	const { key, props, config, base } = data;

	const { variant = "", sizeStyle = "", colorStyle = "" } = props;

	const mode = getMode(props);

	const overrideClassName = config?.overrideClasses?.[key];

	return {
		base,
		overrideClassName,
		applies: [
			config?.styles?.[key],
			config?.modes?.[mode]?.styles?.[key],
			config?.variants?.[variant]?.[key],
			config?.modes?.[mode]?.variants?.[variant]?.[key],
			config?.sizes?.[sizeStyle]?.[key],
			config?.modes?.[mode]?.sizes?.[sizeStyle]?.[key],
			config?.colorStyles?.[colorStyle]?.[key],
			config?.modes?.[mode]?.colorStyles?.[colorStyle]?.[key],
		],
		override: config?.overrides?.[key],
	};
}

export * from "@poodle/system";
