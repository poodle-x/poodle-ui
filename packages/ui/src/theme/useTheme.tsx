import { ThemeContext } from "@emotion/react";
import deepmerge from "deepmerge";
import React from "react";
import { ThemeConfig } from "./";

export function useTheme(
	options: {
		override?: ThemeConfig;
		extend?: ThemeConfig;
	} = {}
): { theme: ThemeConfig } {
	const { override, extend } = options;

	const themeContext: ThemeConfig = React.useContext(ThemeContext);

	const returnTheme = React.useMemo(() => {
		if (override) {
			return override;
		}

		if (extend) {
			return deepmerge(themeContext, extend, {
				arrayMerge: (_, sourceArray) => {
					return sourceArray;
				},
			});
		}

		return themeContext;
	}, [extend, override, themeContext]);

	return {
		theme: returnTheme || {},
	};
}
