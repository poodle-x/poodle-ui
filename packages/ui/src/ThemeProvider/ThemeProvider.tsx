import React from "react";
import { ThemeProvider as BaseThemeProvider } from "@emotion/react";
import { createTheme } from "../theme";
import { PortalContext } from "../Portal";

export interface ThemeProviderProps {
	noDefaultTheme?: boolean;
	theme?: any;
	children?: React.ReactNode;
	disablePortalContext?: boolean;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = (
	props: ThemeProviderProps
) => {
	const [savedPortal, setSavedPortal] = React.useState<
		HTMLElement | undefined
	>();
	const { theme, children, noDefaultTheme, disablePortalContext } = props;

	const memoTheme = React.useMemo(() => {
		if (noDefaultTheme && theme) {
			return theme;
		}

		return createTheme(theme);
	}, [theme, noDefaultTheme]);

	React.useLayoutEffect(() => {
		if (!disablePortalContext) {
			const currentPortal = document.body.querySelector(
				'*[data-portal="poodle-ui"]'
			);
			if (currentPortal) {
				setSavedPortal(currentPortal as HTMLElement);
			} else {
				const newPortal = document.createElement("div");
				newPortal.setAttribute("data-portal", "poodle-ui");
				document.body.appendChild(newPortal);
				setSavedPortal(newPortal);
			}
		}
	}, [disablePortalContext]);

	React.useLayoutEffect(() => {
		if (disablePortalContext && savedPortal) {
			document.body.removeChild(savedPortal);
		}
	}, [disablePortalContext, savedPortal]);

	if (disablePortalContext) {
		return <BaseThemeProvider theme={memoTheme}>{children}</BaseThemeProvider>;
	}

	return (
		<PortalContext.Provider
			value={{
				defaultMount: savedPortal,
			}}
		>
			<BaseThemeProvider theme={memoTheme}>{children}</BaseThemeProvider>
		</PortalContext.Provider>
	);
};

ThemeProvider.displayName = "PoodleThemeProvider";

export default ThemeProvider;
