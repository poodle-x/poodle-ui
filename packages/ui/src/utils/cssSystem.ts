import { CSSObjectSystem } from "../styled";
import { getThemeValue, system, ThemeValue } from "../theme";

export function cssSystem(data: {
	props: any;
	base: CSSObjectSystem[];
	applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
	override?: ThemeValue<CSSObjectSystem>;
}) {
	const { props, applies = [], override, base } = data;

	const overrideTheme = getThemeValue<CSSObjectSystem | undefined>(
		props,
		override,
		undefined
	);

	if (overrideTheme) {
		return system(
			{ ...overrideTheme, theme: props.theme },
			{
				sendProps: props,
			}
		);
	}

	return [
		...base.map((b) =>
			system(
				{ ...b, theme: props.theme },
				{
					sendProps: props,
				}
			)
		),
		...applies.reduce<(CSSObjectSystem | CSSObjectSystem[])[]>((prev, s) => {
			const v = getThemeValue(props, s, undefined);

			if (v) {
				return [
					...prev,
					system({ ...v, theme: props.theme }, { sendProps: props }),
				];
			}

			return prev;
		}, []),
	];
}
