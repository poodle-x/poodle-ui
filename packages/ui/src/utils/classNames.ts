import React from "react";
import { css, CSSObjectSystem, cx } from "../styled";
import { ThemeValue } from "../theme";
import { cssSystem } from "./cssSystem";

type ClassName =
	| undefined
	| string
	| ((props: any) => string)
	| { [key: string]: boolean };

export function genClassNames(data: {
	props: any;
	classNames: Array<ClassName>;
}) {
	const { props, classNames } = data;

	let resultClassNames = "";

	classNames.forEach((cn) => {
		if (typeof cn === "function") {
			resultClassNames = cx(resultClassNames, cn(props));
			return;
		}

		resultClassNames = cx(resultClassNames, cn);
	});

	return resultClassNames;
}

export function classNameCompose(data: {
	props: any;
	overrideClassName?: string | boolean | undefined;
	styles: {
		base: CSSObjectSystem[];
		applies?: (ThemeValue<CSSObjectSystem> | undefined)[];
		override?: ThemeValue<CSSObjectSystem>;
	};
}) {
	const { overrideClassName, styles, props } = data;

	if (typeof overrideClassName === "string") {
		return overrideClassName;
	}

	if (overrideClassName === true) {
		return "";
	}

	return css(
		cssSystem({
			...styles,
			props,
		})
	);
}

export function genListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
): { [key in keyof typeof lists]: string } {
	const result: { [key: string]: string } = {};
	Object.keys(lists).forEach((key) => {
		let resultClassNames = "";
		const listKeyData = lists[key];
		resultClassNames = genClassNames({
			props: { ...baseProps, ...listKeyData.props },
			classNames: listKeyData.classNames,
		});
		result[key] = resultClassNames;
	});

	return result;
}

export function useGenListClassNames(
	baseProps: any,
	lists: {
		[key: string]: {
			props?: any;
			classNames: Array<ClassName>;
		};
	}
) {
	return React.useMemo(() => {
		return genListClassNames(baseProps, lists);
	}, [lists, baseProps]);
}
