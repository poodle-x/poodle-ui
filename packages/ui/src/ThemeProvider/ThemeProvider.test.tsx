import React from "react";
import { ThemeConfig, useTheme, createTheme } from "../theme";
import { render } from "../utils/test";
import ThemeProvider from "./ThemeProvider";
import { PortalContext } from "../Portal";

const TestTheme: React.FC<any> = (props: { keys?: string[] }) => {
	const { keys } = props;
	const { theme } = useTheme();
	if (keys) {
		return JSON.stringify(
			keys.map((k) => {
				return JSON.stringify((theme as any)[k]);
			})
		);
	}

	return JSON.stringify(theme) as any;
};

describe("<ThemeProvider />", () => {
	it("should render children correctly", () => {
		const { container } = render(<ThemeProvider>Hello</ThemeProvider>, {
			withProvider: false,
		});
		expect(container.firstChild).toMatchSnapshot();
	});

	it("disablePortalContext prop should work ", () => {
		function Test() {
			const context = React.useContext(PortalContext);
			return <div>{context ? "With context" : "No context"}</div>;
		}

		const { queryByText, rerender } = render(
			<ThemeProvider>
				<Test />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);

		expect(queryByText("With context")).toBeInTheDocument();

		expect(
			document.body.querySelector('*[data-portal="poodle-ui"]')
		).toBeInTheDocument();

		rerender(
			<ThemeProvider disablePortalContext={true}>
				<Test />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);

		expect(queryByText("No context")).toBeInTheDocument();

		expect(
			document.body.querySelector('*[data-portal="poodle-ui"]')
		).not.toBeInTheDocument();
	});

	it("should render default theme correctly", () => {
		const { container } = render(
			<ThemeProvider>
				<TestTheme />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);
		expect(container.firstChild?.textContent).toEqual(
			JSON.stringify(createTheme())
		);
	});

	it("should extend theme correctly", () => {
		const { container } = render(
			<ThemeProvider
				theme={
					{
						direction: "rtl",
						spacing: {
							scale: 100,
						},
					} as ThemeConfig
				}
			>
				<TestTheme keys={["spacing", "direction"]} />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should render theme correctly with noDefaultTheme option", () => {
		const { container } = render(
			<ThemeProvider theme={{ test: "abc" } as any} noDefaultTheme={true}>
				<TestTheme />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it("should fallback default theme with noDefaultTheme option", () => {
		const resultDefault = render(
			<ThemeProvider>
				<TestTheme />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);

		const resultFallback = render(
			<ThemeProvider noDefaultTheme={true}>
				<TestTheme />
			</ThemeProvider>,
			{
				withProvider: false,
			}
		);

		expect(resultDefault.container.firstChild).toEqual(
			resultFallback.container.firstChild
		);
	});
});
