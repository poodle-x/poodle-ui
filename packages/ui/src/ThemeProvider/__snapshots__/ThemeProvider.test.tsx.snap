// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeProvider /> should extend theme correctly 1`] = `["{\\"scale\\":100,\\"sets\\":{\\"s\\":4,\\"m\\":8,\\"l\\":16}}","\\"rtl\\""]`;

exports[`<ThemeProvider /> should render children correctly 1`] = `Hello`;

exports[`<ThemeProvider /> should render theme correctly with noDefaultTheme option 1`] = `{"test":"abc"}`;
